### ISSUE TITLE
Dark Codex Gather Power changes seems to cause an incompatibility w/ Bubble Buffs.

### ISSUE THREAD

**Issue by Seroster01** [Truinto/DarkCodex#261]
Hello, I'm here to report an issue that results in a mod compatibility issue that (supposedly) originates within Dark Codex; I didn't personally discover the root of the issue, but I was advised by members of the Mod channel on the Owlcat Discord that others had found the root of the issue & after a bit of troubleshooting myself, I have at least some reason to believe they were correct. They advised me to put the bug report here bc they seemed to believe the root of the issue is in Dark Codex rather than the other mod & there doesn't seem to've been an existing bug report for it.

To start with, the incompatibility in question is expressed as such: if a Kineticist is in the party, Bubble Buffs won't populate anything in it's Abilities window except for Treat Affliction. As such, you can't automate any buffs that fall under this section. 

What I've been told, and what I had at least partial evidence for, is that Dark Codex makes some change to Gather Power that sets its duration to 0, which is supposedly an incorrect/incompatible value & this value being 0 somehow results in Bubble Buffs freaking out when it tries to populate the abilities screen. My personal testing showed that removing Gather Power from my kineticist via Toybox caused the issue to (mostly) resolve, but said character also couldn't actually use Kinetic Blasts if Gather Power wasn't in his abilities list, so simply removing the offending ability wasn't an option. 

In any event, I feel hesitant to file this here bc a.) removing all of the DC patches I could see that affected Gather Power didn't seem to resolve the issue & b.) I didn't do the full gruntwork myself & I have little-to-no experience creating mods for WOTR. BUT, on the advice of someone who seemed to be aware of this issue & much more experienced w/ creating mods than me, I decided to file it here. Thank you for your work on this great mod!

**Comment by ScaredKurufinve** [comment]
> that sets its duration to 0, which is supposedly an incorrect/incompatible value & this value being 0 somehow results in Bubble Buffs freaking out when it tries to populate the abilities screen. 

Some clarification: DurationValue on the context action is null, so BubbleBuffs can't check for DurationValue.Rate and throws an NPE.

**Comment by Seroster01** [comment]
After a fair bit of palling around trying to resolve the issue, I'm 99.9% sure that the root issue IS the manual-cast version of Gather Power. Under the proper circumstances, the Abilities window functions normally if I remove it from my character, & when it's on it 100% borks the affected window. Ideally this would make it easy to fix, but there's 2 funky bits to this that have made it difficult to manage: 

1.) the problem ability seems to re-apply itself after every save, even if I disable the Codex patch that's supposed to provide it. If the patch is disabled it doesn't appear in the abilities window, but Toybox still shows it as a removeable ability & that seems to be reflected in the next bit: 

2.) Bearing in mind the fact that it re-asserts itself every time I reload, the BB Abilities window will break if I don't remove the offending ability before opening it the first time after a fresh load, & under this set of circumstances (removed it & subsequently reloaded, causing it to re-apply itself), it appears to stay broken even if I later remove it.

I'm not sure any of this will contribute in any real manner to squashing the bug, but I wanted it known that my testing seems to confirm the root of the issue is the ability in question.

**Comment by Truinto** [comment]
That should be plenty of information. Thanks for the report. I don't
remember changing anything recently and I did use BubbleBuff with a
kineticist the whole time. So not sure what changed since then.

I also never change any duration value on gather power. On first look
all the duration values I create seem to have a non-null value... Very
strange.

Will need to get home before I can dig deeper.

**Comment by Truinto** [comment]
Okay, I think figured it out. The issue is that I don't set the
duration value if I create the object with UseDurationSeconds. If that
flag is used, you can create buffs which last less than 1 round (or 1.5
rounds). That doesn't come up often, but does for gather power. The game
does never read DurationValue, since the flag tells it to read
DurationSeconds. That's why this is totally acceptable from the games
perspective.

But I presume that Bubble always reads DurationValue in some logic (that
was added recently). That's also OK, since the game sets non-null
DurationValue (even tho it wouldn't have to, since UseDurationSeconds
redirects anyway).

In short, I will fix it by adding 1 line of code.

**Comment by Truinto** [comment]
Done. Thanks for all the helpful information. Will update in about 9
hours.

**Comment by Truinto** [comment]
Would make it more robust. But I understand that either approach can be
seen as valid. We have no clear policies on which fields must be
non-null (except for a couple explicit [CanBeNull] [NotNull]
attributes). I already fixed the issue on my end. And I suppose most
modders use BlueprintCore which probably handles that part under the
hood.

**Comment by Seroster01** [comment]
Not trying to rush you, but did this get updated & put in? It sounds like you're saying you already did the fix, but the latest version I'm seeing in Releases is from last week. I'm very noobish at using/navigating github so I'm not exactly sure how updates & the like work; do you update the base files & then package them into a release later? If so, do I have to wait for the updated version to get posted in the Releases section, or is there some other way to get the fix without having to trouble you to make a release version? Thanks again for looking into/fixing this.

**Comment by Seroster01** [comment]
Thanks a bunch!

**Comment by Seroster01** [comment]
Can confirm that the update seems to've fixed the issue; everything appears to be working like normal. Again, thanks a lot for doing this so quickly!
